### PR TITLE
Allow multiple capture groups for debounce regex action

### PR DESCRIPTION
**Pull request by ShivanKaul**
Resolves https://github.com/brave/brave-browser/issues/24750

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/debounce/browser/debounce_rule.cc [hunk 1]
```diff
@@ -187,17 +187,39 @@ bool DebounceRule::ValidateAndParsePatternRegex(
...
   }
 
-  if (!RE2::PartialMatch(path, pattern_regex, parsed_value)) {
+  // Get the number of capturing groups in the expression.
+  size_t number_of_capturing_groups =
+      pattern_regex.NumberOfCapturingGroups() + 1;  
+
+  // Create result vectors to get the matches for the capturing groups.
+  std::vector<re2::StringPiece> match_results(number_of_capturing_groups);
+  re2::StringPiece trailing_slash("/");
```
Comment (fmarier) [inline components/debounce/browser/debounce_rule.cc:202]: That name confused me because it's not just a trailing slash, it's also a path separator.

**Review by fmarier — CHANGES_REQUESTED** [review]
Instead of manually adding slashes (and potentially adding them at the end where it may not work), why not simply capture one of the slashes in a capture group?

e.g. instead of `^/([^/]+)/s/(.*)$`, we could use `^/([^/]+)/s(/.*)$`

**Review by pilgrim-brave — COMMENTED** [review]
Not opposed to the concept. Deferring to François about the regexes. Definitely needs a lot more tests.

File: components/debounce/browser/debounce_rule.cc (hunk 1 above)
Comment (ShivanKaul) [reply components/debounce/browser/debounce_rule.cc:202]: I guess my thought was that we should ensure that the URL is always well-formed, but going back to our earlier discussion, if the regex is wrong we shouldn't compensate for it and just fail.

File: components/debounce/browser/debounce_rule.cc [hunk 2]
```diff
@@ -187,17 +187,35 @@ bool DebounceRule::ValidateAndParsePatternRegex(
...
+  if (!pattern_regex.Match(path, 0, path.size(), RE2::UNANCHORED,
+                           match_results.data(), match_results.size())) {
     VLOG(1) << "Debounce rule with param: " << param_
             << " was unable to capture string";
     return false;
   }
+
+  // Build parsed_value by appending matches, ignoring the first match which
+  // will be the whole match
+  std::for_each(std::begin(match_results) + 1, std::end(match_results),
```
Comment (fmarier) [inline components/debounce/browser/debounce_rule.cc:212]: nit: I think you're already checking for a related failure, but I'd add a `DCHECK(match_results.size() > 1)` just to make it explicit that this line assumes there are at least two elements.

**Review by fmarier — APPROVED** [review]
That looks good to me, but someone more familiar with Chromium idioms should check whether that's the best way to concatenate `StringPiece`s.

**Comment by ShivanKaul** [comment]
> That looks good to me, but someone more familiar with Chromium idioms should check whether that's the best way to concatenate `StringPiece`s.

@bridiver any thoughts? https://github.com/brave/brave-core/pull/14687/files#diff-19bb36033aacc98f6f730c5cbda666206d951a19f782c58fca555b7e44765ab3R215

File: components/debounce/browser/debounce_rule.cc [hunk 3]
```diff
@@ -212,12 +231,14 @@ bool DebounceRule::Apply(const GURL& original_url,
       action_ != kDebounceRegexPath)
     return false;
   // If URL matches an explicitly excluded pattern, this rule does not apply.
-  if (exclude_pattern_set_.MatchesURL(original_url))
+  if (exclude_pattern_set_.MatchesURL(original_url)) {
```
Comment (iefremov) [inline components/debounce/browser/debounce_rule.cc:234]: this diff is not needed, existing formatting is acceptable

**Review by iefremov — APPROVED** [review]

File: components/debounce/browser/debounce_rule.cc (hunk 3 above)
Comment (ShivanKaul) [reply components/debounce/browser/debounce_rule.cc:234]: I know but I just wanted to match everything up. It helps me a tiny bit when I want to add local logging to have these, and given that they're both equivalent I thought might as well. Personally I think always specifying {} is easier to read, but no strong feelings

**Comment by ShivanKaul** [comment]
@MadhaviSeelam [working on adding a test rule](https://github.com/brave/adblock-lists/pull/957) to the debounce.json list, but in the meantime the De-Yandex rules should be sufficient. I've updated the URLs to test in the test plan.

**Comment by ShivanKaul** [comment]
@MadhaviSeelam updated the test plan with the test rule: https://dev-pages.brave.software/navigation-tracking/debouncing.html, all tests should pass (thanks @pes10k!)

**Pull request merged**